src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp

```diff
@@ -206,38 +207,19 @@ static ColumnWithTypeAndName readColumnWithDate64Data(std::shared_ptr<arrow::Chu
 
 static ColumnWithTypeAndName readColumnWithTimestampData(std::shared_ptr<arrow::ChunkedArray> & arrow_column, const String & column_name)
 {
-    auto internal_type = std::make_shared<DataTypeUInt32>();
+    const auto & arrow_type = static_cast<const arrow::TimestampType &>(*(arrow_column->type()));
+    const UInt8 scale = arrow_type.unit() * 3;
+    auto internal_type = std::make_shared<DataTypeDateTime64>(scale, arrow_type.timezone());
     auto internal_column = internal_type->createColumn();
-    auto & column_data = assert_cast<ColumnVector<UInt32> &>(*internal_column).getData();
+    auto & column_data = assert_cast<ColumnDecimal<DateTime64> &>(*internal_column).getData();
     column_data.reserve(arrow_column->length());
 
     for (size_t chunk_i = 0, num_chunks = static_cast<size_t>(arrow_column->num_chunks()); chunk_i < num_chunks; ++chunk_i)
     {
-        auto & chunk = dynamic_cast<arrow::TimestampArray &>(*(arrow_column->chunk(chunk_i)));
-        const auto & type = static_cast<const ::arrow::TimestampType &>(*chunk.type());
-
-        UInt32 divide = 1;
-        const auto unit = type.unit();
-        switch (unit)
-        {
-            case arrow::TimeUnit::SECOND:
-                divide = 1;
-                break;
-            case arrow::TimeUnit::MILLI:
-                divide = 1000;
-                break;
-            case arrow::TimeUnit::MICRO:
-                divide = 1000000;
-                break;
-            case arrow::TimeUnit::NANO:
-                divide = 1000000000;
-                break;
-        }
-
+        const auto & chunk = dynamic_cast<const arrow::TimestampArray &>(*(arrow_column->chunk(chunk_i)));
         for (size_t value_i = 0, length = static_cast<size_t>(chunk.length()); value_i < length; ++value_i)
         {
-            auto timestamp = static_cast<UInt32>(chunk.Value(value_i) / divide); // ms! TODO: check other 's' 'ns' ...
-            column_data.emplace_back(timestamp);
+            column_data.emplace_back(chunk.Value(value_i));
```

Comment:
Potential compatibity issues? I mean it was always casted to DateTime. It seem now it's always DateTime64.

Reply:
But I think it is better to transform timestamp to datetime64 in ch.  The former implementation is not completed, see the TODO..

Reply:
Do you think adding a setting will solve the problem ?

Reply:
> But I think it is better to transform timestamp to datetime64 in ch. The former implementation is not completed, see the TODO..

100% agree, but i guess it may lead to problems during the version upgrade for people who currenly use DateTime64 -> DateTime in Arrow-based file format.

I'm not a big fan of new settings, but we had a precedent, so maybe we should do something similar here too? https://github.com/ClickHouse/ClickHouse/pull/11512/files

Reply:
I guess there will be no problems here. If someone uses DateTime type in CH structure then DateTime64 will be just converted to DateTime with no losses. Like:
```
$ clickhouse client -q "select toDateTime64(100001.1111111111, 9) format Arrow" > data.arrow
$ clickhouse client -q "select * from file(data.arrow, 'Arrow', 'd DateTime64(9)')"
1970-01-02 03:46:41.111111111
$ clickhouse client -q "select * from file(data.arrow, 'Arrow', 'd DateTime')"
1970-01-02 03:46:41 -- works as before
```
The new behaviour could be noticed in schema inference:
```
$ clickhouse client -q "desc file(data.arrow)"
d	DateTime64(9, \'Etc/UTC\')
```
Before it was inferred as `UInt32`, so current behaviour is much better and it should nott lead to compatibility issues